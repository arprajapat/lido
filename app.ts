import { MinHeap } from './MinHeap';
export class TopViewedProduct <T> {
    private hm: Map<T, number>;
    private minHeap: MinHeap<T>;
    constructor(capacity) {
        this.hm = new Map<T, number>();
        this.minHeap = new MinHeap<T>(capacity);
    }

    public processData(arr: T[]) {
        for (let i = 0; i < arr.length; i++) {
            if(!this.hm.has(arr[i])) {
                this.hm.set(arr[i], 0);
            }
            
            this.hm.set(arr[i], this.hm.get(arr[i])+1);
            this.minHeap.add(arr[i], this.hm.get(arr[i]));  
        }
    }

    public getTopViewedProduct (k: number) {
        const arr = this.minHeap.getTopProduct();
        arr.sort((a, b) => b.frequency-a.frequency);
        const min = Math.min(k, arr.length);
        arr.length = min;
        return arr;

    }
}
