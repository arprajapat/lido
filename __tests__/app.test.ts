import { TopViewedProduct } from '../app'

test('Top Viewed Products', () => {
    const topProducts = new TopViewedProduct<number>(5);
    topProducts.processData([1,2,3,4,5,3,4,5,6,6,5,6,5,4,5,6,4,3,2,4,6,6,7,7,7,8]);
    const topThreeProducts = topProducts.getTopViewedProduct(3);
    expect(topThreeProducts).toMatchSnapshot();
    expect(topThreeProducts.length).toEqual(3);
});



