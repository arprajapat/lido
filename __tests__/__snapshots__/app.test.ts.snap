// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Top Viewed Products 1`] = `
Array [
  Object {
    "frequency": 6,
    "key": 6,
  },
  Object {
    "frequency": 5,
    "key": 4,
  },
  Object {
    "frequency": 5,
    "key": 5,
  },
]
`;
